File: python/producer.py
#!/usr/bin/env python3
import pika
import sys
import logging
import json
from datetime import datetime

# Configurar logging
logging.basicConfig(level=logging.INFO, format='%(asctime)s - %(levelname)s - %(message)s')
logger = logging.getLogger(__name__)

def send_message(message):
    """
    Envia uma mensagem para o RabbitMQ
    """
    try:
        # Parâmetros de conexão
        connection_params = pika.ConnectionParameters(
            host='localhost',
            port=5672,
            virtual_host='/',
            credentials=pika.PlainCredentials('guest', 'guest')
        )
        
        # Estabelecer conexão
        connection = pika.BlockingConnection(connection_params)
        channel = connection.channel()
        
        # Configurações da exchange
        exchange_name = 'basic_exchange'
        routing_key = 'basic.message'
        
        # Declarar exchange (caso não exista)
        channel.exchange_declare(
            exchange=exchange_name,
            exchange_type='direct',
            durable=True
        )
        
        # Preparar a mensagem
        message_data = {
            'content': message,
            'timestamp': datetime.now().isoformat(),
            'sender': 'producer'
        }
        
        # Publicar mensagem
        channel.basic_publish(
            exchange=exchange_name,
            routing_key=routing_key,
            body=json.dumps(message_data),
            properties=pika.BasicProperties(
                delivery_mode=2,  # Tornar a mensagem persistente
            )
        )
        
        logger.info(f"Mensagem enviada: {message}")
        
        # Fechar conexão
        connection.close()
        
    except pika.exceptions.AMQPConnectionError:
        logger.error("Erro: Não foi possível conectar ao RabbitMQ. Verifique se o servidor está rodando.")
        sys.exit(1)
    except Exception as e:
        logger.error(f"Erro ao enviar mensagem: {e}")
        sys.exit(1)

if __name__ == '__main__':
    if len(sys.argv) < 2:
        print("Uso: python producer.py 'sua mensagem aqui'")
        sys.exit(1)
    
    message = ' '.join(sys.argv[1:])
    send_message(message)

File: node/producer.js
const amqp = require('amqplib');

// Configurações do RabbitMQ
const RABBITMQ_CONFIG = {
    url: 'amqp://guest:guest@localhost:5672/',
    exchange: 'basic_exchange',
    routingKey: 'basic.message'
};

/**
 * Envia uma mensagem para o RabbitMQ
 * @param {string} message - A mensagem a ser enviada
 */
async function sendMessage(message) {
    let connection;
    let channel;

    try {
        console.log('Conectando ao RabbitMQ...');

        // Estabelecer conexão
        connection = await amqp.connect(RABBITMQ_CONFIG.url);
        channel = await connection.createChannel();

        // Declarar exchange (caso não exista)
        await channel.assertExchange(RABBITMQ_CONFIG.exchange, 'direct', {
            durable: true
        });

        // Preparar a mensagem
        const messageData = {
            content: message,
            timestamp: new Date().toISOString(),
            sender: 'producer'
        };

        const messageBuffer = Buffer.from(JSON.stringify(messageData));

        // Publicar mensagem
        const published = channel.publish(
            RABBITMQ_CONFIG.exchange,
            RABBITMQ_CONFIG.routingKey,
            messageBuffer,
            {
                persistent: true, // Tornar a mensagem persistente
                timestamp: Date.now()
            }
        );

        if (published) {
            console.log(`Mensagem enviada: ${message}`);
        } else {
            console.error('Falha ao enviar mensagem');
        }

    } catch (error) {
        console.error('Erro ao enviar mensagem:', error.message);

        if (error.code === 'ECONNREFUSED') {
            console.error('Erro: Não foi possível conectar ao RabbitMQ. Verifique se o servidor está rodando.');
        }

        process.exit(1);
    } finally {
        // Fechar conexão
        if (channel) {
            await channel.close();
        }
        if (connection) {
            await connection.close();
        }
    }
}

/**
 * Envia múltiplas mensagens para teste
 * @param {Array<string>} messages - Array de mensagens para enviar
 */
async function sendMultipleMessages(messages) {
    for (const message of messages) {
        await sendMessage(message);
        // Pequeno delay entre mensagens
        await new Promise(resolve => setTimeout(resolve, 500));
    }
}

// Processar argumentos da linha de comando
if (require.main === module) {
    const args = process.argv.slice(2);

    if (args.length === 0) {
        console.log('Uso: node producer.js "sua mensagem aqui"');
        console.log('Ou: node producer.js --test (para enviar mensagens de teste)');
        process.exit(1);
    }

    if (args[0] === '--test') {
        const testMessages = [
            'Mensagem de teste 1',
            'Mensagem de teste 2',
            'Mensagem de teste 3',
            'Hello from Node.js!',
            'RabbitMQ funcionando!'
        ];

        console.log('Enviando mensagens de teste...');
        sendMultipleMessages(testMessages);
    } else {
        const message = args.join(' ');
        sendMessage(message);
    }
}

module.exports = { sendMessage, sendMultipleMessages };
